pkg/virt-controller/watch/vmi.go

```diff
@@ -2399,3 +2401,109 @@ func (c *VMIController) syncMemoryHotplug(vmi *virtv1.VirtualMachineInstance) {
 		vmi.Labels[virtv1.MemoryHotplugOverheadRatioLabel] = *overheadRatio
 	}
 }
+
+type aggregateDataVolumeCondition struct {
+	Status  k8sv1.ConditionStatus
+	Reason  []string
+	Message []string
+}
+
+func (c *VMIController) aggregateDataVolumesConditions(vmiCopy *virtv1.VirtualMachineInstance, dvs []*cdiv1.DataVolume) {
+	if len(dvs) == 0 {
+		return
+	}
+	const ConditionUninitialized k8sv1.ConditionStatus = "Uninitialized"
+
+	allDVConditions := map[cdiv1.DataVolumeConditionType]aggregateDataVolumeCondition{
+		cdiv1.DataVolumeReady: {
+			Status: ConditionUninitialized,
+		},
+		cdiv1.DataVolumeBound: {
+			Status: ConditionUninitialized,
+		},
+		cdiv1.DataVolumeRunning: {
+			Status: ConditionUninitialized,
+		},
+	}
+
+	for _, dv := range dvs {
+		for _, condition := range dv.Status.Conditions {
+			if allDVConditions[condition.Type].Status == ConditionUninitialized {
+				allDVConditions[condition.Type] = aggregateDataVolumeCondition{
+					Status:  condition.Status,
+					Reason:  []string{condition.Reason},
+					Message: []string{condition.Message},
+				}
+			} else {
+				if condition.Type == cdiv1.DataVolumeRunning {
+					aggregateCondsTrueWins(condition, allDVConditions)
+				} else {
+					aggregateCondsFalseWins(condition, allDVConditions)
+				}
+			}
+		}
+	}
+
+	dvsReadyCondition := virtv1.VirtualMachineInstanceCondition{
+		Type:   virtv1.VirtualMachineInstanceDataVolumesReady,
+		Status: k8sv1.ConditionUnknown,
+	}
+
+	if allDVConditions[cdiv1.DataVolumeBound].Status == k8sv1.ConditionFalse {
```

Comment:
nit: Maybe make it a switch?

Reply:
Unfortunately I'm checking different values on the various branches.